src/System.Private.CoreLib/shared/System/Attribute.cs

```diff
@@ -0,0 +1,153 @@
+
+                    // The hashcode of an array ignores the contents of the array, so it can produce 
+                    // different hashcodes for arrays with the same contents.
+                    // Since we do deep comparisons of arrays in Equals(), this means Equals and GetHashCode will
+                    // be inconsistent for arrays. Therefore, we ignore hashes of arrays.
+                    if (fieldValue != null && !fieldValue.GetType().IsArray)
+                        vThis = fieldValue;
+
+                    if (vThis != null)
+                        break;
+                }
+
+                if (vThis != null)
+                    return vThis.GetHashCode();
+
+                type = type.BaseType;
+            }
+
+            return type.GetHashCode();
+        }
+#endif
+
+        // Compares values of custom-attribute fields.    
+        private static bool AreFieldValuesEqual(object thisValue, object thatValue)
+        {
+            if (thisValue == null && thatValue == null)
+                return true;
+            if (thisValue == null || thatValue == null)
+                return false;
+
+            Type thisValueType = thisValue.GetType();
+
+            if (thisValueType.IsArray)
+            {
+                // Ensure both are arrays of the same type.
+                if (!thisValueType.Equals(thatValue.GetType()))
+                {
+                    return false;
+                }
+
+                Array thisValueArray = thisValue as Array;
+                Array thatValueArray = thatValue as Array;
+                if (thisValueArray.Length != thatValueArray.Length)
+                {
+                    return false;
+                }
+
+                // Attributes can only contain single-dimension arrays, so we don't need to worry about 
+                // multidimensional arrays.
+                Debug.Assert(thisValueArray.Rank == 1 && thatValueArray.Rank == 1);
+                for (int j = 0; j < thisValueArray.Length; j++)
+                {
+                    if (!AreFieldValuesEqual(thisValueArray.GetValue(j), thatValueArray.GetValue(j)))
+                    {
+                        return false;
+                    }
+                }
+            }
+            else
+            {
+                // An object of type Attribute will cause a stack overflow. 
+                // However, this should never happen because custom attributes cannot contain values other than
+                // constants, single-dimensional arrays and typeof expressions.
+                Debug.Assert(!(thisValue is Attribute));
+                if (!thisValue.Equals(thatValue))
+                    return false;
+            }
+
+            return true;
+        }
+
+        public virtual object TypeId => GetType();
+
+        public virtual bool Match(object obj) => Equals(obj);
+
+        public virtual bool IsDefaultAttribute() => false;
```

Comment:
There is a ton more stuff in `System.Attribute`. Do we really want to only share this much?

Reply:
The rest are static members and will need more refactoring. I wanted to keep the PR simple and will do the statics in a separate pass as it looks like more complex change

Reply:
ok